### DISCUSSION TITLE
type CString# = Addr#

### DISCUSSION THREAD

**Discussion by JakobBruenker** [ghc-proposals/ghc-proposals#499] [Ideas]
I don't use the low-level stuff very much, so I might be completely off-base here, but:

The first time I saw something like `"abc"# :: Addr#`, I was very confused. I thought `"abc"#` was some unique name pointing to an address that might contain anything, but in my particular case happened to store data for a `CString` - when actually, `"abc"#` is effectively an unboxed `CString` literal, and not really a generic address.

This - as well as the types of any functions using `Addr#`s as a string (`unpackCString# :: Addr# -> [Char]` etc.) - would be made a lot clearer, I think, if the types of these literals were called something like `CString` - but `CString` is already taken, by `Ptr CChar`.

However, since `data Ptr a = Ptr Addr#`, any `Ptr` is just a lifted `Addr#`. So, following standard naming conventions, we could make an unlifted version of `CString` and write

```haskell
type CString# = Addr#
```

And then we'd have `"abc"# :: CString#` and `unpackCString# :: CString# -> [Char]`. (Importantly, `:t "abc"#` should also print `CString#`).

Alternatively, we could make it an unlifted newtype, like

```haskell
type CString# :: TYPE AddrRep
newtype CString# = CString# Addr#
```

However, while it would add some type safety, this has the disadvantage of not being fully backwards compatible.

@Ericson2314 points out that being able to keep track of the null-termination invariant is an argument in favor of the `newtype` version.

**Comment by Ericson2314** [comment]
I vote for a `newtype`, as null-termination is a very important invariant for these things!

--------

Should this be for the CLC instead?

**Reply by JakobBruenker** [reply]
Part of it may be, but the type of literals presumably isn't part of the CLC's scope

**Reply by Ericson2314** [reply]
OK, fair point. I guess we might need either a zero-cost `IsString#` that just works on newtypes, or some more special-case magic.

**Reply by JakobBruenker** [reply]
You mean like an analogue of `-XOverloadedStrings`'s `IsString`?

**Reply by Ericson2314** [reply]
Yeah

**Reply by JakobBruenker** [reply]
I can see how that might be desirable, but it's not clear to me that we *need* something like it for this change, rather than just keeping the `"abc"#` literals monomorphic as `CString#`
